### PR TITLE
Oasis contract deploy header

### PR DESCRIPTION
**Pull request by armaniferrante**
TODO:

Once we have the runtime changes ready

- [x] write integration tests with the runtime
- [x] make header validation more general, i.e., address this todo https://github.com/oasislabs/web3c.js/pull/116/files#diff-cf5da82c4e65044bf02f7034a59e0749R131
- [x] check whitelisted keys in the header
- [x] header deploy api for confidential and eth namespaces

### CODE REVIEW

File: test/mockgateway/index.js [hunk 1]
```diff
@@ -115,6 +127,15 @@ async function handleRequest (req) {
   return obj;
 }
 
-module.exports = function () {
-  return http.createServer(onReq);
+function validateHeader(txData) {
+  if (txData.startsWith('00736973002d00001{"expiry":12343333,"confidential":false}')) {
```
Comment (armaniferrante) [inline test/mockgateway/index.js]: TODO: make this validation/parsing more general once we finish the runtime side.

File: web3c/confidential_provider.js [hunk 2]
```diff
@@ -84,6 +88,9 @@ class ConfidentialSendTransform {
     if (!tx.to) {
       // deploy transaction doesn't encrypt anything for v0.5
       tx.data = this._prependConfidential(tx.data);
+      if (tx.header) {
```
Comment (armaniferrante) [inline web3c/confidential_provider.js]: Note that for the first iteration of this, the plan is to **not** rip out the confidential prefix in the transaction data, so that we can minimize runtime changes. So we will have both a confidential prefix and the header option.

File: test/mockgateway/index.js (hunk 1 above)
Comment (willscott) [reply test/mockgateway/index.js]: should this be "if '!' startsWith"?

File: web3c/confidential_provider.js (hunk 2 above)
Comment (willscott) [reply web3c/confidential_provider.js]: should we have a set of known header keys we allow to be serialized?

Comment (armaniferrante) [reply web3c/confidential_provider.js]: Keys for whom?

File: test/mockgateway/index.js (hunk 1 above)
Comment (armaniferrante) [reply test/mockgateway/index.js]: Yes.

Comment (armaniferrante) [reply test/mockgateway/index.js]: Done. The tests were still passing because `0x` wasn't prefixed to the string.

File: web3c/confidential_provider.js (hunk 2 above)
Comment (willscott) [reply web3c/confidential_provider.js]: for key-value pairs. having the library white-list known-recognized keys that will impact the contract, and not allow arbitrary data to be set there.

Comment (armaniferrante) [reply web3c/confidential_provider.js]: Yes agreed. Will wait for the runtime changes to finalize to do this. Adding a todo.

**Comment by armaniferrante** [comment]
The runtime [build](https://buildkite.com/oasislabs/runtime-ethereum/builds/1275#_) now passes with the changes here and in the [e2e tests](https://github.com/oasislabs/e2e-tests/pull/19).

**Comment by armaniferrante** [comment]
@willscott this is now ready for review

File: test/deploy_header.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
+/* globals describe,it */
+const assert = require('assert');
```
Comment (willscott) [inline test/deploy_header.js]: consider also 'chai' or similar if built-in asserts seem awkward

File: test/deploy_header.js [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+const DeployHeader = require('../web3c/deploy_header');
+const DeployHeaderHexReader = DeployHeader.private.DeployHeaderHexReader;
```
Comment (willscott) [inline test/deploy_header.js]: not used in this file?

File: test/web3c.js [hunk 5]
```diff
@@ -4,6 +4,7 @@ const web3 = require('web3');
 const web3c = require('../');
 const gateway = require('./mockgateway');
 const artifact = require('../demo/example.json');
+const responses = require('./mockgateway/responses');
```
Comment (willscott) [inline test/web3c.js]: you both include responses directly, and use the re-exported gateway.responses in this file. choose one.

File: web3c/confidential.js [hunk 6]
```diff
@@ -10,89 +12,107 @@ const Signer = require('./signer');
...
-    }
-
-    c.setProvider.call(this, instanceProvider);
-
-    // Deployed contracts are instantiated with clone.
-    // This patch keeps those clones confidential.
-    this.clone = () => {
-      return new ConfidentialContract(this.options.jsonInterface, this.options.address, this.options);
-    };
+  setupConfidentialRpc(web3) {
```
Comment (willscott) [inline web3c/confidential.js]: the methods as a static method on confidential was setup to parallel the structure used by web3.eth.

I don't think we care about that, especially given the restructuring of that library, but there was reason for why it was structured in a somewhat convoluted way :)

File: web3c/confidential.js [hunk 7]
```diff
@@ -10,89 +12,107 @@ const Signer = require('./signer');
...
+      method.attachToObject(this);
+    });
+  }
+
+  /**
+   * Creates the confidential `Contract` constructor on the `confidential` namespace.
+   * The `Contract` behaves in the same way as the eth `Contract` object, except
+   * transparently encrypts/decrypts all outbound/inbound requests according
+   * to the web3c spec.
+   */
```
Comment (willscott) [inline web3c/confidential.js]: include @params in the doc comment

File: web3c/confidential.js [hunk 8]
```diff
@@ -10,89 +12,107 @@ const Signer = require('./signer');
...
-        }); 
-      });
-    }
-    // deployment
-    tx.data = transformer._prependConfidential(tx.data);
-    return wrappedSigner(tx, from);
-  };
+  }
+
+  setupWrappedAccounts(web3) {
```
Comment (willscott) [inline web3c/confidential.js]: these setup methods are public methods on the confidential object when declared this way. looks like JS doesn't currently have a way to make private class methods. these should be private though...

File: web3c/contract_factory.js [hunk 9]
```diff
@@ -0,0 +1,86 @@
+/**
+ * @returns a constructor to create a web3.eth.Contract object. The `Contract` object
```
Comment (willscott) [inline web3c/contract_factory.js]: @returns goes with @param's after an initial description of the function.

File: web3c/contract_factory.js [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+ *          Oasis contract header upon deploy and the ability to retrieve contract
+ *          expiry from an instance of such a Contract. This is used in both confidential
+ *          and non confidential settings.
+ * @param   {Object} web3 the web3 object.
+ * @param   {Function?} providerFn is a block to execute returning the contract's provider.
+ *          The providerFn should take one argument, the options given upon invoking the
+ *          Contract constructor, for example { saveSession: false } for confidential
+ *          contracts.
+ */
+function make(web3, providerFn) {
```
Comment (willscott) [inline web3c/contract_factory.js]: make isn't a meaningful name for this

File: web3c/contract_factory.js [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+      }],
+      outputFormatter: (res) => res
+    });
+    expiry.setRequestManager(web3._requestManager);
+    expiry.attachToObject(this);
+  }
+
+}
+
+module.exports = {
```
Comment (willscott) [inline web3c/contract_factory.js]: why not set it directly as `module.exports = make;`

File: web3c/contract_factory.js [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+      // Create the txObject that we want to patch and return.
+      let txObject = c.deploy.call(this, deployOptions, callback);
+
+      // Methods we want to hook into.
+      let _send = txObject.send;
+      let _estimateGas = txObject.estimateGas;
+
+      // Perform patches.
+      txObject.send = (options) => {
+        options = Object.assign({}, options);
```
Comment (willscott) [inline web3c/contract_factory.js]: why is this needed?

File: web3c/eth.js [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+  constructor(web3) {
+    Object.assign(this, web3.eth);
+    this.__proto__ = web3.__proto__;
+    this.Contract = ContractFactory.make(web3, (_options) => {
+      return new EthContractProvider(web3._requestManager);
+    });
+  }
+}
+
+class EthContractProvider {
```
Comment (willscott) [inline web3c/eth.js]: are we going to need to wrap accounts on this object as well?

File: web3c/index.js [hunk 14]
```diff
@@ -16,6 +18,8 @@ module.exports = function (provider) {
   localWeb3.call(this, provider);
   if (this.version && !this.version.api) { // v1.0 series
     this.confidential = new Confidential(this, localStorage, MraeBox);
+    this.oasis = new Oasis(this);
```
Comment (willscott) [inline web3c/index.js]: can we have web3.oasis namespace be the non-confidential but with expiry / other oasis-specific functionality namespace?
(i'm wondering if there's a way to avoid monkey-patching the web3.eth namespace entirely)

File: web3c/contract_factory.js (hunk 9 above)
Comment (armaniferrante) [reply web3c/contract_factory.js]: What extra information would be conveyed with such a description?

File: web3c/contract_factory.js (hunk 11 above)
Comment (armaniferrante) [reply web3c/contract_factory.js]: Subjective but I don’t like exporting functions as the module itself. Reads poorly imo.

File: web3c/index.js (hunk 14 above)
Comment (armaniferrante) [reply web3c/index.js]: So we would put the contract object on the oasis namespace if we did this.

File: web3c/eth.js (hunk 13 above)
Comment (armaniferrante) [reply web3c/eth.js]: Good point. We should.

File: web3c/contract_factory.js (hunk 12 above)
Comment (armaniferrante) [reply web3c/contract_factory.js]: In case the deploy options is undefined

File: web3c/contract_factory.js (hunk 11 above)
Comment (armaniferrante) [reply web3c/contract_factory.js]: With that said, it would make more sense to just export the function, since it’s a constructor for a contract factory.

Beats calling this file contract factory factory.

File: web3c/confidential.js (hunk 8 above)
Comment (armaniferrante) [reply web3c/confidential.js]: They should be private, but this structure reads much more nicely imo. Is there a better solution than just adding  _?

File: web3c/confidential.js (hunk 6 above)
Comment (armaniferrante) [reply web3c/confidential.js]: We shouldn’t copy the subjectively annoying syntax of another code base just for the sake of it :P

This way it’s much easier to understand what’s going on  for people not used to js

File: test/deploy_header.js (hunk 3 above)
Comment (armaniferrante) [reply test/deploy_header.js]: We just need this for the rejects method

File: web3c/index.js (hunk 14 above)
Comment (armaniferrante) [reply web3c/index.js]: From an api perspective, I would prefer not doing this. It’s less clear to have three options for contract objects. The eth/confidential dichotomy is obvious for non confidential and confidential.

Adding a third into the mix is not clearly a good idea to me.

Comment (armaniferrante) [reply web3c/index.js]: A more dramatic change would be to drop the eth namespace and just use Oasis. We would still need to do the copying though.

File: test/deploy_header.js (hunk 4 above)
Comment (armaniferrante) [reply test/deploy_header.js]: Removed.

File: test/web3c.js (hunk 5 above)
Comment (armaniferrante) [reply test/web3c.js]: Removed.

File: web3c/confidential.js (hunk 6 above)
Comment (armaniferrante) [reply web3c/confidential.js]: Changed all these methods to `_setup*`.

File: web3c/confidential.js (hunk 7 above)
Comment (armaniferrante) [reply web3c/confidential.js]: Done. Though I would rather have types, e.g. typescript, than non-descriptive comments like this. We could consider when we update to Es6 with @beta38.

File: web3c/contract_factory.js (hunk 10 above)
Comment (armaniferrante) [reply web3c/contract_factory.js]: From the caller's perspective, changed from `ContractFactory.make` to `makeContractFactory`.

File: web3c/contract_factory.js (hunk 11 above)
Comment (armaniferrante) [reply web3c/contract_factory.js]: Done.

File: web3c/contract_factory.js (hunk 12 above)
Comment (armaniferrante) [reply web3c/contract_factory.js]: Let me see if I can remove this line. I recall some weirdness with the monkey patching.

File: web3c/index.js (hunk 14 above)
Comment (willscott) [reply web3c/index.js]: that's what i was suggesting - eth would be the unpatched web3 eth. Oasis would be the patched eth + our options. confidential would be oasis but defaulting to confidential header tagged on.

Comment (armaniferrante) [reply web3c/index.js]: I dig it.

Comment (armaniferrante) [reply web3c/index.js]: Though we then lose the web3 mapping of function to rpc method. E.g., web3.oasis.getBlock => eth_getBlock.

Comment (willscott) [reply web3c/index.js]: we've already lost that with confidential, though 🤔

Comment (armaniferrante) [reply web3c/index.js]: We don't. Only things on confidential are `[ 'keyManager', 'getPublicKey', 'call', 'Contract', 'accounts' ]`.

The only rpcs there are `getPublicKey` and `call`. 

The `Contract` object doesn't map to any rpcs. 1-1 (not even in eth namespace) since it's an object abstraction, not rpc.

File: web3c/eth.js (hunk 13 above)
Comment (armaniferrante) [reply web3c/eth.js]: Done.

File: test/web3c.js [hunk 15]
```diff
@@ -18,15 +18,15 @@ describe('Web3', () => {
     timeout = 5000;
   }
 
-  if (process && process.env && process.env.MNEMONIC) {
+  if (false && process && process.env && process.env.MNEMONIC) {
```
Comment (willscott) [inline test/web3c.js]: these are presumably temporary?

File: web3c/deploy_header.js [hunk 16]
```diff
@@ -0,0 +1,247 @@
...
+
+  /**
+   * @returns the length of the version in nibbles.
+   */
+  static versionLength() {
+    return 2*2;
+  }
+
+  /**
+   * @returns the length, in nibbles, of the size section.
```
Comment (willscott) [inline web3c/deploy_header.js]: the index where the size section starts?

File: test/web3c.js (hunk 15 above)
Comment (armaniferrante) [reply test/web3c.js]: Indeed.

File: web3c/deploy_header.js (hunk 16 above)
Comment (armaniferrante) [reply web3c/deploy_header.js]: Done.

File: web3c/contract_factory.js (hunk 12 above)
Comment (armaniferrante) [reply web3c/contract_factory.js]: So this is needed because `options` is an optional argument. Sometimes it is undefined, in which case we need to replace it with {} so that we can give it the `deployOptions.header` in the next line below.

Comment (willscott) [reply web3c/contract_factory.js]: `options = options || {};`  could be a simpler way to accomplish that?

**Comment by willscott** [comment]
this PR should also bump package version to v2.0.0 as part of it
